theseus/src/event/mod.rs

```diff
@@ -0,0 +1,150 @@
+//! Theseus state management system
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, fmt, path::PathBuf, sync::Arc};
+use tokio::sync::OnceCell;
+use tokio::sync::RwLock;
+use uuid::Uuid;
+
+pub mod emit;
+
+// Global event state
+// Stores the Tauri app handle and other event-related state variables
+static EVENT_STATE: OnceCell<Arc<EventState>> = OnceCell::const_new();
+pub struct EventState {
+    /// Tauri app
+    #[cfg(feature = "tauri")]
+    pub app: tauri::AppHandle,
+    pub loading_bars: RwLock<HashMap<LoadingBarId, LoadingBar>>,
+}
+
+impl EventState {
+    #[cfg(feature = "tauri")]
+    pub async fn init(app: tauri::AppHandle) -> crate::Result<Arc<Self>> {
+        EVENT_STATE
+            .get_or_try_init(|| async {
+                Ok(Arc::new(Self {
+                    app,
+                    loading_bars: RwLock::new(HashMap::new()),
+                }))
+            })
+            .await
+            .map(Arc::clone)
+    }
+
+    #[cfg(not(feature = "tauri"))]
+    pub async fn init() -> crate::Result<Arc<Self>> {
+        EVENT_STATE
+            .get_or_try_init(|| async {
+                Ok(Arc::new(Self {
+                    loading_bars: RwLock::new(HashMap::new()),
+                }))
+            })
+            .await
+            .map(Arc::clone)
+    }
+
+    #[cfg(feature = "tauri")]
+    pub async fn get() -> crate::Result<Arc<Self>> {
+        Ok(EVENT_STATE.get().ok_or(EventError::NotInitialized)?.clone())
+    }
+
+    // Initialization requires no app handle in non-tauri mode, so we can just use the same function
+    #[cfg(not(feature = "tauri"))]
+    pub async fn get() -> crate::Result<Arc<Self>> {
+        Self::init().await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LoadingBar {
+    pub loading_bar_uuid: LoadingBarId,
+    pub message: String,
+    pub total: f64,
+    pub current: f64,
+}
+
+// Loading Bar Id lets us uniquely identify loading bars stored in the state
+// the uuid lets us identify loading bars across threads
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct LoadingBarId {
+    pub key: LoadingBarType,
+    pub uuid: Uuid,
+}
+
+impl LoadingBarId {
+    pub fn new(key: LoadingBarType) -> Self {
+        Self {
+            key,
+            uuid: Uuid::new_v4(),
+        }
+    }
+}
+
+impl fmt::Display for LoadingBarId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}-{}", self.key, self.uuid)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, Copy, PartialEq, Eq, Hash)]
+pub enum LoadingBarType {
+    StateInit,
+    PackDownload,
+    MinecraftDownload,
+    ProfileSync,
```

Comment:
For each of these I'm thinking we add more data along side this, for the PackDownload it shows the name of the pack downloading. For Minecraft the profile ID + name would be cool to have (could just have the whole profile). same with profile sync